trin-core/src/portalnet/find/query_pool.rs

```diff
@@ -0,0 +1,295 @@
+    }
+
+    /// Polls the pool to advance the queries.
+    pub fn poll(&mut self) -> QueryPoolState<'_, TTarget, TNodeId, TResult> {
+        let now = Instant::now();
+        let mut finished = None;
+        let mut waiting = None;
+        let mut timeout = None;
+
+        for (&query_id, query) in self.queries.iter_mut() {
+            query.started = query.started.or(Some(now));
+            match query.next(now) {
+                QueryState::Finished => {
+                    finished = Some(query_id);
+                    break;
+                }
+                QueryState::Waiting(Some(return_peer)) => {
+                    waiting = Some((query_id, return_peer));
+                    break;
+                }
+                QueryState::Waiting(None) | QueryState::WaitingAtCapacity => {
+                    let elapsed = now - query.started.unwrap_or(now);
+                    if elapsed >= self.query_timeout {
+                        timeout = Some(query_id);
+                        break;
+                    }
+                }
+            }
+        }
+
+        if let Some((query_id, return_peer)) = waiting {
+            let query = self.queries.get_mut(&query_id).expect("s.a.");
+            return QueryPoolState::Waiting(Some((query, return_peer)));
+        }
+
+        if let Some(query_id) = finished {
+            let query = self.queries.remove(&query_id).expect("s.a.");
+            return QueryPoolState::Finished(query);
+        }
+
+        if let Some(query_id) = timeout {
+            let query = self.queries.remove(&query_id).expect("s.a.");
+            return QueryPoolState::Timeout(query);
+        }
+
+        if self.queries.is_empty() {
+            QueryPoolState::Idle
+        } else {
+            QueryPoolState::Waiting(None)
+        }
+    }
+}
+
+/// Unique identifier for an active query.
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct QueryId(pub usize);
+
+impl std::ops::Deref for QueryId {
+    type Target = usize;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A query in a `QueryPool`.
+pub struct Query<TTarget, TNodeId, TResult> {
+    /// The unique ID of the query.
+    id: QueryId,
+    /// The peer iterator that drives the query state.
+    peer_iter: QueryPeerIter<TNodeId>,
+    /// The instant when the query started (i.e. began waiting for the first
+    /// result from a peer).
+    started: Option<Instant>,
+    /// Target we are looking for.
+    target: TTarget,
+    /// Phantom use of TResult generic type, which is used in implementation.
+    unused_result: PhantomData<TResult>,
+    
+}
+
+/// The peer selection strategies that can be used by queries.
+enum QueryPeerIter<TNodeId> {
+    FindNode(FindNodeQuery<TNodeId>)
+}
+
+impl<TTarget, TNodeId, TResult> Query<TTarget, TNodeId, TResult>
+where
+    TTarget: TargetKey<TNodeId>,
+    TNodeId: Into<Key<TNodeId>> + Eq + Clone,
+    TResult: Into<TNodeId> + Clone,
+{
```

Comment:
Related to comment below. This implementation seems to imply that the result type will be some key for a node ID. This is also evident in the `on_success` method with the type of the `new_peers` argument.

Reply:
Similar to above, I picture us adding an optional content field to `on_success`.

As for `TResult`, that's a good point, but in my mind I don't see us needing to use `TResult` to return the content if we pass in a channel to send the content back over. In this case, we might be able to get rid of `TResult` altogether. It was mostly used for the predicate queries, which have been removed.

Reply:
Similar to above, I picture us adding an optional content field to on_success.

As for TResult, that's a good point, but in my mind I don't see us needing to use TResult to return the content if we pass in a channel to send the content back over. In this case, we might be able to get rid of TResult altogether. It was mostly used for the predicate queries, which have been removed.